### ISSUE TITLE
Removing the States after logout

### ISSUE THREAD

**Issue by 32teeths** [angular-ui/ui-router#2711]
I am facing this issue mentioned in [page](http://stackoverflow.com/questions/36842688/removing-ui-router-states-dynamically) . I am building states dynamically using $futureStateProvider , and need to remove the states once the user logs out . Or else , the states will be defined again on login , and it breaks . Hoping for a solution. 

THanks

**Comment by eddiemonge** [comment]
The page is inaccessible. Also needs a plnkr example.

**Comment by 32teeths** [comment]
Sorry for mentioning a dead link , The issue is that ,

I have used Future State Provider to build the states dynamically and now trying to remove/reset the states once the user logs out . How do i achieve that ,

**Comment by killerspaz** [comment]
The concept of Providers in Angular is to PREPARE/PROVIDE information PRIOR to bootstrap. Because of this, the idea is it doesn't change once it's been initialized.

There are, however, hooks into ui-router that would work just fine:

$rootScope.$on('$stateChangeStart', function(event, toState, toParams, fromState, fromParams) {
    //Validate user's authentication is still valid

   if (true === isUserAuthenticated && 'login' === toState) {
     //Send user to your primary route since they're already logged in, and attempting to hit the login route.
   }

   if (false === isUserAuthenteicated) {
       event.preventDefault();
       $rootScope.$broadcast('unauthorized'); //or something that indicates to your app that you should go to a login screen.
   }
});

**Comment by 32teeths** [comment]
@killerspaz I am not facing an issue to redirect user to the login page , I need to remove the dynamically created states once the user logs out .

**Comment by christopherthielen** [comment]
I would like to support this in the future, but it's not currently possible to remove state definitions.

**Comment by christopherthielen** [comment]
This is a prerequisite for #2739

**Comment by gochinj** [comment]
I have a mod that adds a removeState method to $stateProvider and a remove method to $urlRoutesProvider that removes the associated rule. My problem is I forked the most recent version from GitHub, but the project won't build.  I have opened an issue waiting for a response. 

In the interim I have just  did a 'copy pasta'   of the distro version of the angular-ui-router.js and made the changes in there.  Once I can build, I will make the changes to the TS files and make a pull request.
